### PR TITLE
Enable the default OpenACC execution space

### PR DESCRIPTION
**Pull request by seyonglee**
Enable the default OpenACC execution space
- Necessary bug fixes and modification to run the incremental tests correctly on the default OpenACC execution space.
- Most of tests are disabled for the OpenACC backend, which will be gradually enabled as more features are implemented.

### CODE REVIEW

File: core/src/Kokkos_Macros.hpp [hunk 1]
```diff
@@ -563,6 +568,29 @@ static constexpr bool kokkos_omp_on_host() { return false; }
...
+#define KOKKOS_IF_ON_DEVICE(CODE)   \
+  if (__builtin_is_device_code()) { \
+    KOKKOS_IMPL_STRIP_PARENS(CODE)  \
+  }
+#define KOKKOS_IF_ON_HOST(CODE)      \
+  if (!__builtin_is_device_code()) { \
+    KOKKOS_IMPL_STRIP_PARENS(CODE)   \
+  }
+#else
+// FIXE acc_on_device is a non-constexpr function
```
Comment (dalg24) [inline core/src/Kokkos_Macros.hpp:582]:
```suggestion
// FIXME_OPENACC acc_on_device is a non-constexpr function
```

File: core/unit_test/TestInit.hpp [hunk 2]
```diff
@@ -70,4 +70,7 @@ TEST(TEST_CATEGORY, dispatch) { test_dispatch<TEST_EXECSPACE>(); }
 }  // namespace Test
 
 #include <TestCompilerMacros.hpp>
+#ifndef KOKKOS_ENABLE_OPENACC
+// FIXME_OPENACC: not implemented yet.
```
Comment (dalg24) [inline core/unit_test/TestInit.hpp:74]:
```suggestion
#ifndef KOKKOS_ENABLE_OPENACC. // FIXME_OPENACC not implemented
```
put on a single line
same comment everywhere

Comment (dalg24) [reply core/unit_test/TestInit.hpp:74]: Actually in this case say "TeamPolicy not implement"

File: core/unit_test/TestViewAPI.hpp [hunk 3]
```diff
@@ -1386,13 +1386,17 @@ class TestViewAPI {
       Kokkos::deep_copy(h, d);
     }
 
+#if !defined(KOKKOS_ENABLE_OPENACC)
+    // FIXME_OPENACC: this can be removed when MDRange parallel_for is
+    // implemented.
```
Comment (dalg24) [inline core/unit_test/TestViewAPI.hpp:1391]:
```suggestion
#if !defined(KOKKOS_ENABLE_OPENACC)  // FIXME_OPENACC MDRange parallel_for not implemented
```

**Review by dalg24 — CHANGES_REQUESTED** [review]
Put FIXME_OPENACC and short explanatory comment on the same line.
It comes handy when you grep for them

**Comment by dalg24** [comment]
I meant for you to fix all occurrences

**Comment by seyonglee** [comment]
> I meant for you to fix all occurrences

Updated the comments.

File: core/src/OpenACC/Kokkos_OpenACC_DeepCopy.hpp [hunk 4]
```diff
@@ -88,8 +88,7 @@ struct Kokkos::Impl::DeepCopy<Kokkos::Experimental::OpenACCSpace,
         "Kokkos::Impl::DeepCopy<OpenACCSpace, OpenACCSpace, "
         "ExecutionSpace>::DeepCopy: fence before copy");
     if (n > 0) {
-      acc_memcpy_device_async(dst, const_cast<void*>(src), n,
-                              exec.acc_async_queue());
+      acc_memcpy_device(dst, const_cast<void*>(src), n);
```
Comment (dalg24) [inline core/src/OpenACC/Kokkos_OpenACC_DeepCopy.hpp:91]: Why did you change?

File: core/unit_test/TestInit.hpp [hunk 5]
```diff
@@ -70,4 +70,6 @@ TEST(TEST_CATEGORY, dispatch) { test_dispatch<TEST_EXECSPACE>(); }
 }  // namespace Test
 
 #include <TestCompilerMacros.hpp>
+#ifndef KOKKOS_ENABLE_OPENACC  // FIXME_OPENACC not implemented
```
Comment (dalg24) [inline core/unit_test/TestInit.hpp:73]: Say it is missing "TeamPolicy"

**Review by dalg24 — COMMENTED** [review]
Mostly OK except for the changes to `DeepCopy`

File: core/src/OpenACC/Kokkos_OpenACC_DeepCopy.hpp (hunk 4 above)
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_DeepCopy.hpp:91]: This case is for non-OpenACC execution space (`exec`) to execute the memory copy (`acc_memcpy_device`), where the `exec` does not have `acc_async_queue` method, and thus synchronous memory copy should happen.
One possible alternative implementation is to assign a default async queue for the OpenACC backend, and this default async queue is used for non-OpenACC execution space to perform asynchronous memory transfer.

File: core/unit_test/TestInit.hpp (hunk 5 above)
Comment (seyonglee) [reply core/unit_test/TestInit.hpp]: Updated.

File: core/src/OpenACC/Kokkos_OpenACC_DeepCopy.hpp (hunk 4 above)
Comment (dalg24) [reply core/src/OpenACC/Kokkos_OpenACC_DeepCopy.hpp:91]: Oh right

**Review by dalg24 — APPROVED** [review]

**Review by Rombur — APPROVED** [review]
